File: server/java/src/main/java/com/onlinepayments/example/api/validators/common/iban/IbanValidator.java
package com.onlinepayments.example.api.validators.common.iban;

import com.onlinepayments.example.api.exceptions.ApiValidationException;
import com.onlinepayments.example.service.domain.common.enums.Country;

import static com.onlinepayments.example.api.validators.BaseValidator.isNullOrEmpty;

public class IbanValidator {
    public static void validateIban(String iban, Country country) {
        if (isNullOrEmpty(iban)) {
            return;
        }

        String cleanedIban = iban.replaceAll("[^A-Za-z0-9]", "").toUpperCase();

        if (!cleanedIban.matches("^[A-Z]{2}\\d{2}[A-Z0-9]+$")) {
            throw new ApiValidationException(
                    "The IBAN filed format is invalid (expected: 2 letters country + 2 digits + alphanumerics)."
            );
        }

        if (country != null) {
            String prefix;
            int expectedLength;

            switch (country) {
                case France -> {
                    prefix = "FR";
                    expectedLength = 27;
                }
                case Germany -> {
                    prefix = "DE";
                    expectedLength = 22;
                }
                case England -> {
                    prefix = "GB";
                    expectedLength = 22;
                }
                default -> throw new ApiValidationException("The IBAN country is not supported.");
            }

            if (!cleanedIban.startsWith(prefix) || cleanedIban.length() != expectedLength) {
                throw new ApiValidationException(
                        "The IBAN field must start with '" + prefix + "' and be " +
                                expectedLength + " characters for " + country.name() + "."
                );
            }
        }

        if (!hasValidIbanChecksum(cleanedIban)) {
            throw new ApiValidationException("IBAN checksum is invalid.");
        }
    }

    private static boolean hasValidIbanChecksum(String iban) {
        String rearranged = iban.substring(4) + iban.substring(0, 4);

        int remainder = 0;
        for (int i = 0; i < rearranged.length(); i++) {
            char ch = rearranged.charAt(i);
            String digits = Character.isLetter(ch)
                    ? String.valueOf((ch - 'A') + 10)
                    : String.valueOf(ch);

            for (char d : digits.toCharArray()) {
                remainder = (remainder * 10 + (d - '0')) % 97;
            }
        }

        return remainder == 1;
    }
}

File: server/python/app/presentation/validators/common/iban/iban_validator.py
import re
from app.application.domain.enums.common.country import Country

BASIC_IBAN_REGEX = re.compile(r'^[A-Z]{2}\d{2}[A-Z0-9]+$')
IBAN_LENGTHS = {"FR": 27, "DE": 22, "GB": 22}

class IbanValidator:

    def validate(self, iban: str, country: str = None) -> list[str]:
        errors = []
        if not iban or not iban.strip():
            return errors

        if not re.match(r'^[a-zA-Z0-9\s]+$', iban):
            errors.append("IBAN format is invalid (expected: 2 letters country + 2 digits + alphanumerics).")

            return errors

        normalized = iban.replace(" ", "").upper()

        if not BASIC_IBAN_REGEX.match(normalized):
            errors.append("IBAN format is invalid (expected: 2 letters country + 2 digits + alphanumerics).")

            return errors

        if not self._has_valid_checksum(normalized):
            errors.append("IBAN checksum is invalid.")

            return errors

        if country:
            iso = Country.to_iso_alpha2(country)
            if iso not in IBAN_LENGTHS:
                errors.append("IBAN country is not supported.")
            else:
                expected_length = IBAN_LENGTHS[iso]
                if not normalized.startswith(iso) or len(normalized) != expected_length:
                    errors.append(f"IBAN must start with '{iso}' and be {expected_length} characters for {country}.")

        return errors

    def _has_valid_checksum(self, iban: str) -> bool:
        try:
            rear = iban[4:] + iban[:4]
            numeric = ""
            for ch in rear:
                if ch.isalpha():
                    numeric += str(ord(ch) - ord('A') + 10)
                elif ch.isdigit():
                    numeric += ch
                else:
                    return False
            remainder = 0
            for i in range(0, len(numeric), 7):
                part = str(remainder) + numeric[i:i+7]
                remainder = int(part) % 97

            return remainder == 1
        except Exception:
            return False